### PR TITLE
Plugins: Fs: Add option to access unallowed files in dev mode

### LINKED ISSUE #66472: Bug: Plugins: 404 when accessing newly created plugin asset files until Grafana is restarted
**What happened**:

It's not possible to access newly created plugin asset files from `/public/plugins` until Grafana is restarted since Grafana 9.5.0. This is especially useful when developing plugins.

**What you expected to happen**:

The new files should be accessible from `/public/plugins` like it was in 9.4.0, at least when running Grafana in development mode.
(However, some people forget to run Grafana in development mode when developing plugins, so it would be better to restore the previous behaviour if possible)

**How to reproduce it (as minimally and precisely as possible)**:

- Install a plugin and run Grafana
- Try to access `/public/plugins/X/module.js` -> the file is returned correctly
- Create a new file "test.txt" inside the plugin's folder, alongside `module.js`
- Try to access `/public/plugins/X/test.txt` -> a 404 is returned:

```json
{
  "message": "Plugin file not found"
}
```

- Restart Grafana
- Try to access `/public/plugins/X/test.txt` -> the file is returned correctly

**Anything else we need to know?**:

**Environment**:
- Grafana version: broken 9.5.0 / main branch, working in 9.4.x
- Data source type & version:
- OS Grafana is installed on: Linux / OS X
- User OS & Browser: Linux / OS X
- Grafana plugins:
- Others:

### PR DESCRIPTION
**Pull request by xnyo**
**What is this feature?**

(see #66472  for more info)

The tl;dr is:
- We have introduced a side effect with the LocalFS refactoring where newly created files are not returned after the plugin is loaded
- This is great for security, but it's not great for development, as it would requires to restart Grafana each time new files are added as part of developing a plugin

This PR:
- Adds a `plugins.StaticFS`, which wraps a `plugins.FS`, and acts as an "allow-list" checked before accessing the file
- The `StaticFS` is added on top of the existing `FS` only when running Grafana in production mode
- Also does some refactoring to `plugins.FS` and `plugins.LocalFS` 

**Why do we need this feature?**

Restores the old behaviour in dev mode (good for plugin development) while also keeping security high for production mode (with the allow-list)

**Who is this feature for?**

Plugin developers and Grafana users in general

**Which issue(s) does this PR fix?**:

Fixes #66472 

**Special notes for your reviewer:**

A lot of changed lines are either comments to `plugins.FS` and test changes.

Please check that:
- [x] It works as expected from a user's perspective.
- [x] If this is a pre-GA feature, it is behind a feature toggle.
- [x] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: pkg/plugins/manager/loader/finder/local.go [hunk 1]
```diff
@@ -24,15 +25,21 @@ var (
 )
 
 type Local struct {
-	log log.Logger
+	log     log.Logger
+	devMode bool
```
Comment (wbrowne) [inline pkg/plugins/manager/loader/finder/local.go:29]: I think it would be better if we just had the cfg as a field here instead. Passing just a boolean to the constructor gives me immediate [Uncle Bob nightmares](https://www.informit.com/articles/article.aspx?p=1392524) 😄

File: pkg/plugins/manager/loader/loader_test.go [hunk 2]
```diff
@@ -32,7 +32,7 @@ var localFSComparer = cmp.Comparer(func(fs1 plugins.LocalFS, fs2 plugins.LocalFS
 	fs1Files := fs1.Files()
 	fs2Files := fs2.Files()
 
-	finder.NewLocalFinder()
+	finder.NewLocalFinder(false)
```
Comment (wbrowne) [inline pkg/plugins/manager/loader/loader_test.go:35]: I guess this whole line can be deleted? 🤔

File: pkg/plugins/localfiles.go [hunk 3]
```diff
@@ -10,32 +10,49 @@ import (
 	"github.com/grafana/grafana/pkg/util"
 )
 
-var _ fs.FS = (*LocalFS)(nil)
+var _ FS = (*LocalFS)(nil)
 
 // LocalFS is a plugins.FS that allows accessing files on the local file system.
 type LocalFS struct {
```
Comment (wbrowne) [inline pkg/plugins/localfiles.go:16]: I wonder if it might be better to have a separate implementation for this instead of adding this logic to the existing one 🤔 Maybe that's too weird though cc @andresmgot @marefr WDYT?

My thinking is that the existing one is quite locked down as it takes the list of files it reads on startup, but a development mode implementation should allow anything within the plugin directory - better not to mix the two concerns.

Comment (marefr) [reply pkg/plugins/localfiles.go:16]: Yeah maybe. Tricky. Wasn't really aware (forgot) of the allow list in the `LocalFS`. I guess that's security related, but we also have this https://github.com/grafana/grafana/blob/8ab7ba2b30ee793ea8d9df5a87810edf02887394/pkg/plugins/manager/signature/manifest.go#L196-L220 that wouldn't even load the plugin if it fails. For plugins in development mode I guess there's no manifest at all and therefore it just worked before?

Is the allowlist in `LocalFS` for the scenario where Grafana has loaded the plugin and verified the manifest and someone sneaks extra files into the plugin dir?

Maybe wrap the `LocalFS` in another `FS`, e.g. an `AllowFS`?

Comment (xnyo) [reply pkg/plugins/localfiles.go:16]: > that wouldn't even load the plugin if it fails. For plugins in development mode I guess there's no manifest at all and therefore it just worked before?

Yes, exactly. It would cause the signature check to fail if there are extra files, but in dev mode unsigned plugins are allowed.

> Is the allowlist in LocalFS for the scenario where Grafana has loaded the plugin and verified the manifest and someone sneaks extra files into the plugin dir?

I think the allow list was just a side effect of Will's refactoring around FS abstraction, but it is good to have to protect against the case you mentioned. I like it as it provides additional security, but it slows things down in develpoment mode because it forces the developer to restart Grafana if a file is added as part of "hot-reloading" a frontend plugin while developing it (LocalFS with allow list is also used to serve plugin assets under /public/plugins/PLUGIN_ID/xyz). Esteban reported it and he was really confused by this behaviour.

> Maybe wrap the LocalFS in another FS, e.g. an AllowFS?

That's a good idea. I wanted to do it like this originally but then switched to options-based as both use a "local" filesystem, so share most of the implementation, but I like it either ways so happy to switch 😄

File: pkg/plugins/localfiles.go [hunk 4]
```diff
@@ -10,79 +11,189 @@ import (
...
+	}
+}
+
+// emptyCollectFilesFunc is a no-op CollectFilesFunc. It returns a filepath.WalkFunc that does nothing and
+// the accumulator map is not altered in any way. This is suitable to use with plugins.allowListFSNoFiles for tests.
+func emptyCollectFilesFunc(_ map[string]struct{}) filepath.WalkFunc {
+	return func(path string, info fs.FileInfo, err error) error {
+		return nil
+	}
+}
```
Comment (andresmgot) [inline pkg/plugins/localfiles.go:39]: can you move this and `NewAllowListLocalFSForTests` to a different file? (to avoid mixing test and functional logic). No need to use a different package, a `test_utils.go` should work.

**Review by wbrowne — COMMENTED** [review]
Just a couple initial things. Feels a bit more complicated / more changes than expected, especially for the sake of something that's only dev mode - but seems to make sense at least

File: pkg/plugins/localfiles.go [hunk 5]
```diff
@@ -10,79 +11,189 @@ import (
...
+	collectFilesFunc CollectFilesFunc
+}
 
-	return LocalFS{
-		m:        pfs,
-		basePath: basePath,
+// NewLocalFS returns a new LocalFS that can access any file in the specified base path on the filesystem.
+// basePath must use os-specific path separator for Open() to work properly.
+// If collectFilesFunc is not provided, defaultCollectFilesFunc is used instead, which will collect all files.
+func NewLocalFS(basePath string, collectFilesFunc CollectFilesFunc) LocalFS {
```
Comment (andresmgot) [inline pkg/plugins/localfiles.go:53]: as I understand, specifying `collectFilesFunc` is only useful for tests. To make that clear / easier to use, can you remove the param from this function and expose a different one `NewFakeLocalFS` that receives the collect function?

File: pkg/plugins/localfiles.go [hunk 6]
```diff
@@ -10,79 +11,189 @@ import (
...
+// The provided name must be a relative file name that uses os-specific path separators.
+// The function returns the corresponding fs.File for accessing the file on the local filesystem.
+// If a nil error is returned, the caller should take care of calling Close() the returned fs.File.
+// If the file does not exist, ErrFileNotExist is returned.
 func (f LocalFS) Open(name string) (fs.File, error) {
 	cleanPath, err := util.CleanRelativePath(name)
-	if err != nil {
-		return nil, err
+	if strings.Contains(cleanPath, "..") || err != nil {
+		return nil, ErrFileNotExist
```
Comment (andresmgot) [inline pkg/plugins/localfiles.go:68]: should this error be something different than "not exists"?

**Review by andresmgot — COMMENTED** [review]
Just some minor comments from my side :+1:

File: pkg/plugins/manager/loader/finder/local.go [hunk 7]
```diff
@@ -81,15 +88,23 @@ func (l *Local) Find(ctx context.Context, src plugins.PluginSource) ([]*plugins.
...
+		if !l.devMode {
+			// In prod, tighten up security by allowing access only to the files present up to this point.
+			// Any new file "sneaked in" won't be allowed and will acts as if the file did not exist.
+			// Build the allow-list by calling Files() once
+			fsFiles, err := pluginFs.Files()
+			if err != nil {
+				return nil, err
+			}
+			// Wrap FS with allow list
+			pluginFs = plugins.NewAllowListFS(pluginFs, fsFiles...)
```
[the remark this answers is not in the document]
Comment (xnyo) [reply pkg/plugins/manager/loader/finder/local.go:102]: The idea was to have `AllowListFS` to act as an additioanl "filter" in front of any `plugins.FS`, so you could technically give it an underlying `LocalFS` (which can access a whole folder on the local filesystem), but by wrapping it into an `AllowListFS` you could allow access to a subset of files/folders exposed by the `LocalFS`.

I think we can either:

1. Add a `NewAutoAllowListFS` or something similar, which will build the allow-list based on the result of `Files()` on creation, and it will keep working in this case.

```go
func NewAutoAllowListFS(fs FS) (AllowListFS, error) {
	fsFiles, err := pluginFs.Files()
	if err != nil {
		return nil, err
	}
	return NewAllowListFS(fs, fsFiles...)
}
```

2. ...or change `NewAllowListFS` to always have that behaviour (removing the possibility to act as a "filter", but just avoiding allowing access to files added after initialization).

Any preferences or alternative solutions? 2 is fine for this case, 1 is more flexible but not sure if we need such flexibility now, as we only have LocalFS and CDN (which acts a bit differently anyways).

File: pkg/plugins/localfiles.go (hunk 6 above)
Comment (xnyo) [reply pkg/plugins/localfiles.go:68]: Ops, good point, it should have been:

```go
if err != nil {
	return nil, err
}
if strings.Contains(cleanPath, "..") {
	return nil, ErrFileNotExist
}
```

I'll change that :grinning:

File: pkg/plugins/manager/loader/finder/local.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (wbrowne) [reply pkg/plugins/manager/loader/finder/local.go:102]: Yeah I think option 2 makes sense for now 👍

File: pkg/plugins/localfiles.go (hunk 6 above)
Comment (andresmgot) [reply pkg/plugins/localfiles.go:68]: well, true, but I meant that having '..' in the path mean that the path is invalid not that it doesn't exist, no?

File: pkg/plugins/localfiles.go [hunk 8]
```diff
@@ -10,79 +11,189 @@ import (
...
+		if strings.Contains(clenRelPath, "..") || err != nil {
+			continue
+		}
+		relFiles = append(relFiles, clenRelPath)
+	}
+	return relFiles, nil
+}
+
+// fsAllowList is a set-like map that contains files that can be accessed from a plugins.FS.
+type fsAllowList map[string]struct{}
```
[the remark this answers is not in the document]
Comment (xnyo) [reply pkg/plugins/localfiles.go:111]: :+1: Makes sense especially if we go with option 2 above, will address shortly!

File: pkg/plugins/manager/loader/finder/local.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (xnyo) [reply pkg/plugins/manager/loader/finder/local.go:102]: Thanks, I'll add that shortly :smile:

**Comment by xnyo** [comment]
Addressed some changes and simplified things a bit, however I am reverting back to draft for now as I have to fix tests and tidy a couple of things up

File: pkg/plugins/localfiles.go (hunk 6 above)
Comment (xnyo) [reply pkg/plugins/localfiles.go:117]: God rid of this error case in the latest version as it was pretty much already handled by `CleanRelativePath`

File: pkg/plugins/localfiles.go [hunk 9]
```diff
@@ -10,76 +11,153 @@ import (
...
+// NewLocalFS returns a new LocalFS that can access any file in the specified base path on the filesystem.
+// basePath must use os-specific path separator for Open() to work properly.
+func NewLocalFS(basePath string) LocalFS {
+	return LocalFS{basePath: basePath}
+}
+
+// fileIsAllowed takes an absolute path to a file and an os.FileInfo for that file, and it checks if access to that
+// file is allowed or not. Access to a file is allowed if the file is in the FS's Base() directory, and if it's a
+// symbolic link it should not end up outside the plugin's directory.
+func (f LocalFS) fileIsAllowed(basePath string, absolutePath string, info os.FileInfo) (bool, error) {
```
Comment (xnyo) [inline pkg/plugins/localfiles.go:36]: To simplify things, I have moved `collectFilesWithin` that we use with the finder/loader into here and removed the ability to specify custom filepath.WalkerFunc on a LocalFS (from a previous iteration of this PR), because we don't really need that functionality for now

**Review by marefr — APPROVED** [review]
LGTM I think this looks good :+1: But had some hard time following all the changes in localfiles.go so would like @wbrowne and @andresmgot to review as well.

File: pkg/plugins/localfiles.go [hunk 10]
```diff
@@ -89,10 +162,83 @@ func (f LocalFS) Remove() error {
...
+	}
+	// Use the wrapped FS to access the file
+	return f.FS.Open(name)
+}
+
+// Files returns a slice of all the file paths in the FS relative to the base path.
+// It calls Files() on the underlying FS, and intersects the result with the allow-list.
+func (f StaticFS) Files() ([]string, error) {
+	// Get files from the underlying FS
+	filesystemFiles, err := f.FS.Files()
```
Comment (wbrowne) [inline pkg/plugins/localfiles.go:227]: Couldn't we skip this and just return what's inside the `f.staticFilesMap`?

Comment (xnyo) [reply pkg/plugins/localfiles.go:227]: Good point. You're right, this can be changed due to the other changes done as part of the PR review

**Review by andresmgot — APPROVED** [review]
Chages LGTM :+1:

**Review by andresmgot — APPROVED** [review]

**Comment by bassmaster187** [comment]
Hi, how can I allow every new (jpg) file in my plugin?

**Comment by wbrowne** [comment]
@bassmaster187 Currently it's not possible as we introduced this for improved security. The development mode version will permit new files so that when developing their plugins, developers won't experience issues with assets 404ing. 

Is it the case that your plugin is creating new files (thumbnail images) on the fly? In that case, do you not also experience issues with signature validation since they are not covered in the signature?

**Comment by bassmaster187** [comment]
> Is it the case that your plugin is creating new files (thumbnail images) on the fly? In that case, do you not also experience issues with signature validation since they are not covered in the signature?

I skip signature validation for my Plugin.
The images are created from another c# application and placed into the plugin folder.
So the only solution would be to restart the whole grafana server after an image has been created?

image

**Comment by xnyo** [comment]
Hi,

If your plugin is unsigned, (private and not intended to be published on the Grafana plguin catalog) you can probably achieve a similar result by implementing a [CallResource handler](https://github.com/grafana/grafana-plugin-sdk-go/blob/d8f8ae671a289d2d077195c8c603ec4485628b66/backend/resource.go#L103) that returns files from the disk.

There's also [an adapter](https://github.com/grafana/grafana-plugin-sdk-go/blob/d8f8ae671a289d2d077195c8c603ec4485628b66/backend/resource/httpadapter/handler.go) that allows you to use implement an HTTP server on top of `CallResource`, allowing you to return images.

We have some examples of this in our [grafana-plugin-examples repo](https://github.com/grafana/grafana-plugin-examples) like here:

- CallResource HTTP adapter and mux definition: https://github.com/grafana/grafana-plugin-examples/blob/e71964e633e19ee193d020f16df117638b7dcce3/examples/app-with-backend/pkg/plugin/app.go#L30-L35

- HTTP routes definition: https://github.com/grafana/grafana-plugin-examples/blob/e71964e633e19ee193d020f16df117638b7dcce3/examples/app-with-backend/pkg/plugin/resources.go

- Calling HTTP resources from the frontend: https://github.com/grafana/grafana-plugin-examples/blob/e71964e633e19ee193d020f16df117638b7dcce3/examples/app-with-backend/src/pages/PageOne/PageOne.tsx#L12

**Please keep in mind that, for security reasons, we do not allow such usage for plugins intended for the catalog and for installation on Grafana Cloud, so a similar approach is viable only for private and unsigned plugins.**
